In [1]:
import pandas as pd
df = pd.read_csv("../../main/h&m_dataset/customer_hm.csv")
df.head()

customer_id  FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...   0       0   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...   0       0   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   0       0   
3  00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...   0       0   
4  00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...   1       1   

  club_member_status fashion_news_frequency  age  
0             ACTIVE                   NONE   49  
1             ACTIVE                   NONE   25  
2             ACTIVE                   NONE   24  
3             ACTIVE                   NONE   54  
4             ACTIVE              Regularly   52

In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 6 columns):
 #   Column                  Non-Null Count    Dtype
---  ------                  --------------    -----
 0   customer_id             1048575 non-null  str  
 1   FN                      1048575 non-null  int64
 2   Active                  1048575 non-null  int64
 3   club_member_status      1048575 non-null  str  
 4   fashion_news_frequency  1048574 non-null  str  
 5   age                     1048575 non-null  int64
dtypes: int64(3), str(3)
memory usage: 48.0 MB


In [3]:
df.shape

(1048575, 6)

In [4]:
df.isna().sum()

customer_id               0
FN                        0
Active                    0
club_member_status        0
fashion_news_frequency    1
age                       0
dtype: int64

6개 컬럼에 대해 총 1048576행의 데이터가 있는데, fashion_news_frequency의 경우 하나의 행만 결측치가 있음  
→ 전체 행 수에 비해 결측치 데이터가 매우 적은 케이스이므로 해당 행을 제거하고 분석을 진행해도 무방하다고 판단

### fashion_news_frequency 결측치 제거

In [5]:
df_drop = df.dropna(subset=['fashion_news_frequency'])

In [6]:
df_drop.info()

<class 'pandas.DataFrame'>
Index: 1048574 entries, 0 to 1048574
Data columns (total 6 columns):
 #   Column                  Non-Null Count    Dtype
---  ------                  --------------    -----
 0   customer_id             1048574 non-null  str  
 1   FN                      1048574 non-null  int64
 2   Active                  1048574 non-null  int64
 3   club_member_status      1048574 non-null  str  
 4   fashion_news_frequency  1048574 non-null  str  
 5   age                     1048574 non-null  int64
dtypes: int64(3), str(3)
memory usage: 56.0 MB


In [7]:
df_drop.isna().sum()

customer_id               0
FN                        0
Active                    0
club_member_status        0
fashion_news_frequency    0
age                       0
dtype: int64

결측치 제거 완료해서 기존 1048575 → 제거 후 1048574 (행 수 ▼1)

### age 이상치 제거

In [8]:
df_drop.head()

customer_id  FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...   0       0   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...   0       0   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   0       0   
3  00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...   0       0   
4  00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...   1       1   

  club_member_status fashion_news_frequency  age  
0             ACTIVE                   NONE   49  
1             ACTIVE                   NONE   25  
2             ACTIVE                   NONE   24  
3             ACTIVE                   NONE   54  
4             ACTIVE              Regularly   52

In [9]:
df_drop.describe().round(3)

FN       Active          age
count  1048574.000  1048574.000  1048574.000
mean         0.356        0.346       36.369
std          0.479        0.476       14.309
min          0.000        0.000       16.000
25%          0.000        0.000       24.000
50%          0.000        0.000       32.000
75%          1.000        1.000       49.000
max          1.000        1.000       99.000

약 36% 정도가 Fashion News 구독  
약 35% 정도가 Fashion News 구독 & Active  

In [10]:
# age가 80 이상인 데이터의 건수 합계
count_80 = len(df_drop[df_drop['age'] >= 80])

print(f"80세 이상 데이터 건수: {count_80}건")

80세 이상 데이터 건수: 1147건


In [11]:
total_count = len(df_drop)
percentage = (count_80 / total_count) * 100

print(f"80세 이상 비중: {percentage:.4f}%")

80세 이상 비중: 0.1094%


**Age의 Min: 16, Max: 99**  
**→ 80세 이상 데이터는 전체 데이터의 0.1% 수준이고, 너무 고령이라 분석에서 유의미한 인사이트를 도출하기에는 필요하지 않다고 판단하여 제거**

In [12]:
df_drop = df_drop[df_drop['age'] < 80]

In [13]:
df_drop.describe().round(3)

FN       Active          age
count  1047427.000  1047427.000  1047427.000
mean         0.355        0.346       36.318
std          0.479        0.476       14.233
min          0.000        0.000       16.000
25%          0.000        0.000       24.000
50%          0.000        0.000       32.000
75%          1.000        1.000       49.000
max          1.000        1.000       79.000

**고령 고객 제거 후:**  
**Age의 Min: 16, Max: 79**  

In [14]:
df_drop.info()

<class 'pandas.DataFrame'>
Index: 1047427 entries, 0 to 1048574
Data columns (total 6 columns):
 #   Column                  Non-Null Count    Dtype
---  ------                  --------------    -----
 0   customer_id             1047427 non-null  str  
 1   FN                      1047427 non-null  int64
 2   Active                  1047427 non-null  int64
 3   club_member_status      1047427 non-null  str  
 4   fashion_news_frequency  1047427 non-null  str  
 5   age                     1047427 non-null  int64
dtypes: int64(3), str(3)
memory usage: 55.9 MB


In [15]:
df_drop.isna().sum()

customer_id               0
FN                        0
Active                    0
club_member_status        0
fashion_news_frequency    0
age                       0
dtype: int64

### 나이 그룹 분류
10대/20대/30대/40대/50대/60대+(60대~70대)

In [16]:
# 나이를 10단위로 나누어 연령대 컬럼 생성
df_drop['age_group'] = (df_drop['age'] // 10) * 10

# 연령대별 고객 분포 확인
print(df_drop['age_group'].value_counts().sort_index())

age_group
10     55256
20    409196
30    181111
40    157445
50    174715
60     56124
70     13580
Name: count, dtype: int64


60대와 70대를 합쳐도 약 7만건 정도라서 10대 제외 나머지 연령대랑 비교하면 숫자가 적은 편임  
**→ 60대 이상으로 한번에 그룹 짓기**

In [17]:
# 60세 이상은 모두 60으로 통일 (60대 이상 그룹화)
df_drop['age_group'] = df_drop['age'].apply(lambda x: 60 if x >= 60 else (x // 10) * 10)

# 결과 확인
print(df_drop['age_group'].value_counts().sort_index())

age_group
10     55256
20    409196
30    181111
40    157445
50    174715
60     69704
Name: count, dtype: int64


### 각 컬럼의 고유값 확인

In [18]:
df_drop.head()

customer_id  FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...   0       0   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...   0       0   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   0       0   
3  00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...   0       0   
4  00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...   1       1   

  club_member_status fashion_news_frequency  age  age_group  
0             ACTIVE                   NONE   49         40  
1             ACTIVE                   NONE   25         20  
2             ACTIVE                   NONE   24         20  
3             ACTIVE                   NONE   54         50  
4             ACTIVE              Regularly   52         50

In [19]:
df_drop['FN'].value_counts()

FN
0    675069
1    372358
Name: count, dtype: int64

In [20]:
df_drop['Active'].value_counts()

Active
0    684910
1    362517
Name: count, dtype: int64

In [21]:
df_drop['club_member_status'].value_counts()

club_member_status
ACTIVE        981597
PRE-CREATE     65471
LEFT CLUB        359
Name: count, dtype: int64

In [22]:
df_drop['fashion_news_frequency'].value_counts()

fashion_news_frequency
NONE         674002
Regularly    372767
Monthly         658
Name: count, dtype: int64

### 컬럼 고유값 형식 동일하게 변경(ex. ACTIVE → Active)

In [23]:
new_df = df_drop.copy()

In [24]:
new_df['club_member_status'] = df_drop['club_member_status'].str.title()
new_df['club_member_status'].value_counts()

club_member_status
Active        981597
Pre-Create     65471
Left Club        359
Name: count, dtype: int64

In [25]:
new_df.head()

customer_id  FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...   0       0   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...   0       0   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   0       0   
3  00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...   0       0   
4  00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...   1       1   

  club_member_status fashion_news_frequency  age  age_group  
0             Active                   NONE   49         40  
1             Active                   NONE   25         20  
2             Active                   NONE   24         20  
3             Active                   NONE   54         50  
4             Active              Regularly   52         50

In [26]:
new_df['fashion_news_frequency'] = df_drop['fashion_news_frequency'].str.capitalize()
new_df['fashion_news_frequency'].value_counts()

fashion_news_frequency
None         674002
Regularly    372767
Monthly         658
Name: count, dtype: int64

### FN 값과 fashion_news_frequency 간 관계 파악

**FN**  
**0    675069**  
1    372358  
Name: count, dtype: int64  

**fashion_news_frequency**  
**None         674002**  
Regularly    372767  
Monthly         658  
Name: count, dtype: int64  

→ FN에서 0 개수와 fashion_news_frequency에서 None 개수가 맞지 않음

차이가 발생하게 된 이유 추측:  
FN(뉴스 구독)은 0으로 해놓곤, 뉴스 수신 빈도는 Regularly 또는 Monthly로 설정  
**→ 어쨌든 0으로 설정해뒀으면 뉴스 구독 안하는거니까 FN==0 이라면 fashion_news_frequency도 None으로 변경해줘야 함**

In [27]:
new_df.loc[new_df['FN'] == 0, 'fashion_news_frequency'] = 'None'

In [28]:
new_df['fashion_news_frequency'].value_counts()

fashion_news_frequency
None         675693
Regularly    371084
Monthly         650
Name: count, dtype: int64

In [29]:
new_df['FN'].value_counts()

FN
0    675069
1    372358
Name: count, dtype: int64

그런데 왜 FN 0인 행 수보다 fashion_news_frequency None인 행 수가 더 많을까?  
→ FN 1인데 주기 설정을 아직 안해서 뉴스 못 받는 경우도 존재

In [30]:
# FN이 1인데도 frequency가 'None'인 사람들의 수
mismatch_count = len(new_df[(new_df['FN'] == 1) & (new_df['fashion_news_frequency'] == 'None')])

print(f"수신동의는 했지만 설정은 None인 사람: {mismatch_count}명")

수신동의는 했지만 설정은 None인 사람: 624명


쇼핑 뉴스를 받은 사람과 받지 않은 사람을 확실히 구분하기 위해서  
위 624명에 대해 FN 0으로 수정한다

In [31]:
new_df.loc[(new_df['FN'] == 1) & (new_df['fashion_news_frequency'] == 'None'), 'FN'] = 0

In [32]:
new_df['fashion_news_frequency'].value_counts()

fashion_news_frequency
None         675693
Regularly    371084
Monthly         650
Name: count, dtype: int64

In [33]:
new_df['FN'].value_counts()

FN
0    675693
1    371734
Name: count, dtype: int64

**이제 FN이 0이라면 fashion_news_frequency도 None으로 맞춰진 상태**

In [34]:
new_df.head()

customer_id  FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...   0       0   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...   0       0   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   0       0   
3  00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...   0       0   
4  00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...   1       1   

  club_member_status fashion_news_frequency  age  age_group  
0             Active                   None   49         40  
1             Active                   None   25         20  
2             Active                   None   24         20  
3             Active                   None   54         50  
4             Active              Regularly   52         50

In [35]:
new_df = new_df.rename(columns={
    'customer_id': 'Customer_Id',
    'club_member_status': 'Club_Member_Status',
    'fashion_news_frequency': 'Fashion_News_Frequency',
    'age': 'Age',
    'age_group': 'Age_Group'
})
new_df.head()

Customer_Id  FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...   0       0   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...   0       0   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   0       0   
3  00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...   0       0   
4  00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...   1       1   

  Club_Member_Status Fashion_News_Frequency  Age  Age_Group  
0             Active                   None   49         40  
1             Active                   None   25         20  
2             Active                   None   24         20  
3             Active                   None   54         50  
4             Active              Regularly   52         50

In [36]:
new_df.shape

(1047427, 7)

In [37]:
new_df.to_csv('./data/cleaned_customer_hm.csv', index = False)

In [38]:
new_df.info()

<class 'pandas.DataFrame'>
Index: 1047427 entries, 0 to 1048574
Data columns (total 7 columns):
 #   Column                  Non-Null Count    Dtype
---  ------                  --------------    -----
 0   Customer_Id             1047427 non-null  str  
 1   FN                      1047427 non-null  int64
 2   Active                  1047427 non-null  int64
 3   Club_Member_Status      1047427 non-null  str  
 4   Fashion_News_Frequency  1047427 non-null  str  
 5   Age                     1047427 non-null  int64
 6   Age_Group               1047427 non-null  int64
dtypes: int64(4), str(3)
memory usage: 63.9 MB
